### ISSUE TITLE
Rule to enforce usage of `from __future__ import annotations` when imports could be moved to type-checking block ?

### ISSUE THREAD

**Issue by Greesb** [astral-sh/ruff#11103]
Hello wonderful developers !

I was wondering if it would be possible to implement a rule to raise an error when an import could be moved to a type-checking block if `from __future__ import annotations` was used ?

Here's a really sparse example:
```python
import datetime

def printme(d: datetime.datetime) -> None:
    print(d.isoformat())
```

Here, `TCH003` doesn't say anything because `datetime` is used for typing.
But if we add `from __future__ import annotations` then the `TCH003` rule triggers and says that `datetime` import should be moved to a type-checking block.

What do you think ?

**Comment by AlexWaygood** [comment]
Hi! I think you can do this by enabling our [I002 rule](https://docs.astral.sh/ruff/rules/missing-required-import/), and using this setting in your `pyproject.toml` file:

```toml
[tool.ruff.lint.isort]
required-imports = ["from __future__ import annotations"]
```

You also might find our [`flake8-future-annotations` rules](https://docs.astral.sh/ruff/rules/#flake8-future-annotations-fa) useful.

Do either/both of those work for you? :-)

**Comment by Greesb** [comment]
Might work as a hack yeah, but then we'll have those imports on file when it might not be needed. And it'll require a lot of ignores because of imports used only for pydantic validation for example.

**Comment by charliermarsh** [comment]
> And it'll require a lot of ignores, because of imports inside custom types used only for pydantic or sqlalchemy validation for example, which still need to be imported at runtime.

Adding the `from __future__ import annotations` on its own will not cause problems here -- you'd only experience issues if you then moved imports into the type-checking block that were needed in such cases. We do have automated rules for that (https://docs.astral.sh/ruff/rules/#flake8-type-checking-tch), and have settings to mark modules like Pydantic that need to be available at runtime:

- https://docs.astral.sh/ruff/settings/#lint_flake8-type-checking_runtime-evaluated-base-classes
- https://docs.astral.sh/ruff/settings/#lint_flake8-type-checking_runtime-evaluated-decorators

So I think `required-imports = ["from __future__ import annotations"]` and the flake8-type-checking rules do what you want.

**Comment by charliermarsh** [comment]
You can also just use the flake8-type-checking rules with [`quote-annotations`](https://docs.astral.sh/ruff/settings/#lint_flake8-type-checking_quote-annotations), and the usages will be quoted to allow imports to be moved into type-checking blocks.

**Comment by Greesb** [comment]
Managed to get something working with your recommendations, thank you !
